storage/database/src/key.rs

```diff
@@ -44,11 +76,41 @@ pub trait Key: Storable + Sized {
 }
 
 //---------------------------------------------------------------------------------------------------- Impl
-impl<T> Key for T
-where
-    T: Storable + Sized,
-{
-    type Primary = Self;
+/// [`Ord`] comparison for arrays/vectors.
+impl<const N: usize, T> Key for [T; N] where T: Key + Storable + Sized + bytemuck::Pod {}
+impl<T: bytemuck::Pod + Debug> Key for StorableVec<T> {}
+
+/// [`Ord`] comparison for any `T`.
+///
+/// This is not a blanket implementation because
+/// it allows outer crates to define their own
+/// comparison functions for their `T: Storable` types.
+impl Key for () {}
+impl Key for StorableBytes {}
+impl Key for StorableStr {}
+
+/// Integer comparison for numbers.
+macro_rules! impl_key_ne_bytes {
+    ($($t:ident),* $(,)?) => {
+        $(
+            impl Key for $t {
+                #[inline]
+                fn compare(left: &[u8], right: &[u8]) -> Ordering {
+                    // INVARIANT:
+                    // This is native endian since [`Storable`] (bytemuck, really)
+                    // (de)serializes bytes with native endian.
+                    let left = $t::from_ne_bytes(left.try_into().unwrap());
+                    let right = $t::from_ne_bytes(right.try_into().unwrap());
+                    std::cmp::Ord::cmp(&left, &right)
+                }
+            }
+        )*
+    };
+}
+
+impl_key_ne_bytes! {
+    u8,u16,u32,u64,u128,usize,
+    i8,i16,i32,i64,i128,isize,
 }
```

Comment:
## Key 4/4
`Key`'s blanket impl is removed and replaced with some implementation for common types.

These impls now implement the behavior that `cuprate_blockchain` wants by default.

| Common type | Comparison behavior |
|-------------|---------------------|
| Vectors/arrays/string | Straight byte comparison; `a.cmp(&b)`
| Numbers | Convert to number, then do a _value_ comparison

The reason why strings are treated the same as bytes is because [Rust does it like that](https://doc.rust-lang.org/1.79.0/src/core/str/traits.rs.html#12-25). We can change it but I don't think we have any tables that care about this.

The reason why the blanket impl is removed is it allows outer crates to define their own
comparison functions for their `T: Storable` types.

The reverse way of doing this is:
- Leaving the blanket impl
- Doing byte comparisons for everything by default

Then in `cuprate_blockchain` we would have to create wrapper types for all the numbers, e.g. `struct SortU8(pub u8)` and implement the custom sorting. This is probably more clean and separated but kind of a pain.

It's worth noting in `redb`:
- [Number comparison is the default](https://docs.rs/redb/2.1.1/src/redb/types.rs.html#601-624)
- [Key types are selectively implemented, not blanket impl on all values](https://docs.rs/redb/2.1.1/redb/trait.Key.html)

So I think doing it like this is OK.

Reply:
Yeah it's probably better to do it this new way